source/sets_functions_and_relations.rst

```diff
@@ -1293,4 +550,457 @@ at an appropriate point in the proof.
 ..   cases h i,
 ..   { assumption },
 ..   contradiction
-.. end
\ No newline at end of file
+.. end
+
+
+.. _functions:
+
+Functions
+---------
+
+In Lean, a function ``f : α → β`` is a function between types.
+If ``p`` is a set of elements of type ``β``,
+the library defines ``preimage f p``, written ``f ⁻¹' p``,
+to be ``{ x | f x ∈ p }``.
+The expression ``x ∈ f ⁻¹' p`` reduces to ``f x ∈ s``.
+This is often convenient, as in the following example:
+
+.. code-block:: lean
+
+    import data.set.function
+
+    variables {α β : Type*}
+    variable  f : α → β
+    variables u v : set β
+
+    example : f ⁻¹' (u ∩ v) = f ⁻¹' u ∩ f ⁻¹' v :=
+    by { ext, refl }
+
+
+If ``s`` is a set of elements of type ``α``,
+the library also defines ``image f s``,
+written ``f '' s``,
+to be ``{y | ∃ x, x ∈ s ∧ f x = y}``.
+So a hypothesis  ``y ∈ f '' s`` decomposes to a triple
+``⟨x, xs, xeq⟩`` with ``x : α`` satisfying the hypotheses ``xs : x ∈ s``
+and ``xeq : f x = y``.
+The ``rfl`` tag in the ``rintros`` tactic (see :numref:`the_existential_quantifier`) was made precisely
+for this sort of situation.
+
+.. code-block:: lean
+
+    import data.set.function
+
+    variables {α β : Type*}
+    variable  f : α → β
+    variables s t : set α
+
+    -- BEGIN
+    example : f '' (s ∪ t) = f '' s ∪ f '' t :=
+    begin
+      ext y, split,
+      { rintros ⟨x, xs | xt, rfl⟩,
+        { left, use [x, xs] },
+        right, use [x, xt] },
+      rintros (⟨x, xs, rfl⟩ | ⟨x, xt, rfl⟩),
+      { use [x, or.inl xs] },
+      use [x, or.inr xt]
+    end
+    -- END
+
+Notice also that the ``use`` tactic applies ``refl``
+to close goals when it can.
+
+Here is another example:
+
+.. code-block:: lean
+
+    import data.set.function
+
+    variables {α β : Type*}
+    variable  f : α → β
+    variables s t : set α
+
+    -- BEGIN
+    example : s ⊆ f ⁻¹' (f '' s) :=
+    begin
+      intros x xs,
+      show f x ∈ f '' s,
+      use [x, xs]
+    end
+    -- END
+
+We can replace the line ``use [x, xs]`` by
+``apply mem_image_of_mem f xs`` if we want to
+use a theorem specifically designed for that purpose.
+But knowing that the image is defined in terms
+of an existential quantifier is often convenient.
```

Comment:
Good idea. I just added the following paragraph:

"The following equivalence is a good exercise:

.. code-block:: lean

    import data.set.function

    variables {α β : Type*}
    variable  f : α → β
    variables s t : set α

    -- BEGIN
    example : f '' s ⊆ t ↔ s ⊆ f ⁻¹' t :=
    sorry
    -- END

It shows that ``image f`` and ``preimage f`` are
an instance of what is known as a *Galois connection*
between ``set α`` and ``set β``,
each partially ordered by the subset relation.
In the library, this equivalence is named
``image_subset_iff``.
In practice, the right-hand side is often the
more useful representation,
because ``y ∈ f ⁻¹' t`` unfolds to ``f y ∈ t``
whereas working with ``x ∈ f '' s`` requires
decomposing an existential quantifier."

I was planning to have a section in this chapter called "Relations." I was going to talk about equivalence relations and use the example of quotienting a preorder, but now I think that maybe the quotient construction should go in a different chapter, with other type constructions. Maybe we can write a good "Relations" section on Galois connections? If you send me a sketch or an outline, I'll be happy to fill it out.